packages/x-date-pickers/src/internals/hooks/useField/useFieldState.ts

```diff
@@ -209,6 +199,55 @@ export const useFieldState = <
     [state.sections],
   );
 
+  // Keep invalid state "sticky" while sections still represent an invalid date.
+  const hasInvalidSectionValue = React.useMemo(() => {
+    if (state.sections.every((s) => s.value === '')) {
+      return false;
+    }
+
+    const allFilled = state.sections.every((s) => s.value !== '');
+    if (!allFilled) {
+      return false;
+    }
+
+    if (activeSectionIndex == null) {
+      return false;
+    }
+
+    const activeDateSections = fieldValueManager.getDateSectionsFromValue(
+      state.sections,
+      state.sections[activeSectionIndex] as any,
+    );
+
+    const dateFromSections = getDateFromDateSections(
+      adapter,
+      activeDateSections as any,
+      localizedDigits,
+    );
+    return !adapter.isValid(dateFromSections);
+  }, [adapter, fieldValueManager, state.sections, activeSectionIndex, localizedDigits]);
+
+  // When the field loses focus (no active section), consider partially filled sections as invalid.
+  // This enforces that the field must be entirely filled or entirely empty on blur.
+  const hasPartiallyFilledSectionsOnBlur = React.useMemo(() => {
+    if (activeSectionIndex != null) {
+      return false;
+    }
+
+    const someFilled = state.sections.some((s) => s.value !== '');
+    const someEmpty = state.sections.some((s) => s.value === '');
```

Comment:
performance wise, better to create a single for loop.

Reply:
You can count the amount of filled value and do both computations from here.